nomad/structs/csi.go

```diff
@@ -707,6 +724,126 @@ func (v *CSIVolume) Validate() error {
+func (v *CSIVolume) Merge(other *CSIVolume) error {
+	if other == nil {
+		return nil
+	}
+
+	var errs *multierror.Error
+
+	if v.Name != other.Name && other.Name != "" {
+		errs = multierror.Append(errs, errors.New("volume name cannot be updated"))
+	}
+	if v.ExternalID != other.ExternalID && other.ExternalID != "" {
+		errs = multierror.Append(errs, errors.New(
+			"volume external ID cannot be updated"))
+	}
+	if v.PluginID != other.PluginID {
+		errs = multierror.Append(errs, errors.New(
+			"volume plugin ID cannot be updated"))
+	}
+	if v.CloneID != other.CloneID && other.CloneID != "" {
+		errs = multierror.Append(errs, errors.New(
+			"volume clone ID cannot be updated"))
+	}
+	if v.SnapshotID != other.SnapshotID && other.SnapshotID != "" {
+		errs = multierror.Append(errs, errors.New(
+			"volume snapshot ID cannot be updated"))
+	}
+
+	// must be compatible with capacity range
+	// TODO: when ExpandVolume is implemented we'll need to update
+	// this logic https://github.com/hashicorp/nomad/issues/10324
+	if v.Capacity != 0 {
+		if other.RequestedCapacityMax < v.Capacity ||
+			other.RequestedCapacityMin > v.Capacity {
+			errs = multierror.Append(errs, errors.New(
+				"volume requested capacity update was not compatible with existing capacity"))
+		} else {
+			v.RequestedCapacityMin = other.RequestedCapacityMin
+			v.RequestedCapacityMax = other.RequestedCapacityMax
+		}
+	}
+
+	// must be compatible with volume_capabilities
+	if v.AccessMode != CSIVolumeAccessModeUnknown ||
+		v.AttachmentMode != CSIVolumeAttachmentModeUnknown {
+		var ok bool
+		for _, cap := range other.RequestedCapabilities {
+			if cap.AccessMode == v.AccessMode &&
+				cap.AttachmentMode == v.AttachmentMode {
+				ok = true
+				break
+			}
+		}
+		if ok {
+			v.RequestedCapabilities = other.RequestedCapabilities
+		} else {
+			errs = multierror.Append(errs, errors.New(
+				"volume requested capabilities update was not compatible with existing capability in use"))
+		}
+	} else {
+		v.RequestedCapabilities = other.RequestedCapabilities
+	}
+
+	// topologies are immutable, so topology request changes must be
+	// compatible with the existing topology, if any
+	if len(v.Topologies) > 0 {
+		if other.RequestedTopologies == nil {
+			if v.RequestedTopologies != nil {
+				errs = multierror.Append(errs, errors.New(
+					"volume topology request update was not compatible with existing topology"))
+			}
+		} else {
+			var err error
+			for _, otherTopo := range other.RequestedTopologies.Required {
+				if !otherTopo.MatchFound(v.Topologies) {
+					err = errors.New(
+						"volume topology requirement update was not compatible with existing topology")
+					break
+				}
+			}
+			for _, otherTopo := range other.RequestedTopologies.Preferred {
+				if !otherTopo.MatchFound(v.Topologies) {
+					err = errors.New(
+						"volume topology preference update was not compatible with existing topology")
+					break
+				}
+			}
```

Comment:
If topologies are immutable, do we need to check in the other direction as well? Meaning, can an update request _reduce_ the scope of the topologies?

I wrote a test case like this and it fails because it doesn't return an error"
```go
		{
			name: "invalid topology update - missing",
			v: &CSIVolume{
				Topologies: []*CSITopology{
					{Segments: map[string]string{"rack": "R1"}},
					{Segments: map[string]string{"rack": "R3"}},
				},
			},
			update: &CSIVolume{
				RequestedTopologies: &CSITopologyRequest{
					Required: []*CSITopology{
						{Segments: map[string]string{"rack": "R1"}},
					},
				},
			},
			expected: "volume topology request update was not compatible with existing topology",
			expectFn: func(t *testing.T, v *CSIVolume) {
				require.Len(t, v.Topologies, 2)
			},
		},
```

Reply:
I think we're ok, but I should _definitely_ add a test covering this case that asserts the expected behavior so that it's not only "accidentally ok". 

The `TopologyRequirement` message in the [CreateVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume) spec has examples, but the high-level description is this:

>  If requisite is specified, the provisioned volume MUST be accessible from at least one of the requisite topologies.

The easiest example is to assume we're creating an AWS EBS volume. These are accessible within single availability zone. So suppose I provide the following topology request:

```hcl
topology_request {
  required {
    topology { segments { "topology.ebs.csi.aws.com/zone" = "us-east-1a" } }
    topology { segments { "topology.ebs.csi.aws.com/zone" = "us-east-1b" } }
  }
}
```

The plugin can create the volume in _either_ `us-east-1a` or `us-east-1b`, and the resulting volume object topology will have topology like this:

```go
&CSIVolume{
	Topologies: []*CSITopology{
		{Segments: map[string]string{"topology.ebs.csi.aws.com/zone": "us-east-1a"},
		},
	},
}
```

So what we're trying to do here is to allow  user to update their request to be:


```hcl
topology_request {
  required {
    topology { segments { "topology.ebs.csi.aws.com/zone" = "us-east-1a" } }
  }
}
```

Because that request could have been sent and resulted in the same topology.

All _that_ being said, having described all that I'm wondering whether it really makes sense to let users change the `topology_request` at all if it can't result in a changed topology. This feels like something we'll write in the docs and then folks will open issues about anyways because it's confusing behavior.

Reply:
Hm, that behavior is even more confusing because you _can't_ reduce your request to:

```
topology_request {}
```

I think I'm going to change this so that once `Topologies` is set by the plugin, you can't change the `TopologyRequest` at all. That's going to have less weird edge cases.